### ISSUE TITLE
Develop a framework to handle provider attributes

### ISSUE THREAD

**Issue by paigejulianne** [genderdirectory/genderdirectory#9]
Providers attach multiple "attributes" (category, payment methods, insurance accepted, etc.) to their master records.  A series of classes should be written to handle these attributes and how they are shown or entered into the system.

For example, the attribute "category" may only have a single textual value entered through a simple text input box.  Furthermore, we may have validation requirements on that attribute and only allow one value to be selected.  The attribute "insurance" may need to hold multiple textual inputs.  We may also have an attribute "lgbtq_identified" that's a simple yes/no option.

The first thing to develop is a series of classes/subclasses of a generic type, for example, `PropGeneric_SingleTextBox` or `PropGeneric_YesNo.`  That could then be extended to handle the actual attribute, for example, `PropHandler_LgbtqIdentified.`  The classes would handle showing, hiding, editing, or searching that attribute.  So, we may have something that looks like:

```
// these would exist for every kind of input or output type. example - Yes/No Toggle, Text Input, Dropdown, Textarea, custom widget we build later
public class PropGeneric_YesNo {

   public method getEdit(): string {
      $returnData = '<label for=":uniqueId">:displayLabel</label><select id=":uniqueId">';
      $returnData .= '<option value="1">Yes</option><option value="0">No</option>
      $returnData .= '</select>';
      $returnData = Helper::substituteStringVariables(['uniqueId' => Helper::randomString(10), 'displayLabel' => $this->getLabel, 'value' => $this->getValueDisplay()]);
      return $returnData;
   }

   public method getValueDisplay(): string {
       return ($this->value) ? 'Yes' : 'No';
   }


}


public class PropHandler_LgbtqIdentified extends PropGeneric_YesNo {

  public method getLabel(): string {
     return 'Is the provider LGBTQIA2S+ identified?';
   }
}

```
Example of database table:

| id | provider_id | attribute | value |
|--------|--------|--------|--------|
| 1 | 1 | Category | General Practitioner |
| 2 | 1 | InsuranceAccepted | Aetna |
| 3 | 1 | InsuranceAccepted | Cigna |
| 4 | 1 | LgbtqIdentified | 1 |

_Use comments below to ask for clarifications or otherwise discuss this plan_

**Comment by paigejulianne** [comment]
**Seeking additional help in the design and implementation of this CORE feature**

**Comment by CedricTiako** [comment]
In order for me to provide you with a more sophisticated rendering, I need more explanation of your processes. 
`// Base generic class for a generic attribute
abstract class PropGeneric {

    protected $value;
    protected $label;
    protected $uniqueId;

    public function __construct($value, $label) {
        $this->value = $value;
        $this->label = $label;
        $this->uniqueId = $this->generateUniqueId();
    }

    protected function generateUniqueId() {
        return uniqid();
    }

    public function getLabel() {
        return $this->label;
    }

    public function getUniqueId() {
        return $this->uniqueId;
    }

    abstract public function getEdit(): string;

    abstract public function getValueDisplay(): string;
}

// Generic class for single text input
class PropGeneric_SingleTextBox extends PropGeneric {

    public function getEdit(): string {
        $editHtml = '<label for="' . $this->uniqueId . '">' . $this->label . '</label>';
        $editHtml .= '<input type="text" id="' . $this->uniqueId . '" value="' . htmlspecialchars($this->value) . '">';
        return $editHtml;
    }

    public function getValueDisplay(): string {
        return htmlspecialchars($this->value);
    }
}

// Generic class for yes/no selection
class PropGeneric_YesNo extends PropGeneric {

    public function getEdit(): string {
        $editHtml = '<label for="' . $this->uniqueId . '">' . $this->label . '</label>';
        $editHtml .= '<select id="' . $this->uniqueId . '">';
        $editHtml .= '<option value="1"' . ($this->value ? ' selected' : '') . '>Yes</option>';
        $editHtml .= '<option value="0"' . (!$this->value ? ' selected' : '') . '>No</option>';
        $editHtml .= '</select>';
        return $editHtml;
    }

    public function getValueDisplay(): string {
        return $this->value ? 'Yes' : 'No';
    }
}

// Specific class for the LGBTQ+ identified attribute
class PropHandler_LgbtqIdentified extends PropGeneric_YesNo {

    public function __construct($value) {
        parent::__construct($value, 'Is the provider LGBTQIA2S+ identified?');
    }
}

// Specific class for the Category attribute
class PropHandler_Category extends PropGeneric_SingleTextBox {

    public function __construct($value) {
        parent::__construct($value, 'Category');
    }

    public function validate(): bool {
        // Specific validation for category
        return !empty($this->value) && is_string($this->value);
    }
}

// Generic class for attributes with multiple values (e.g., insurances)
class PropGeneric_MultiTextBox extends PropGeneric {

    public function getEdit(): string {
        $editHtml = '<label for="' . $this->uniqueId . '">' . $this->label . '</label>';
        $editHtml .= '<textarea id="' . $this->uniqueId . '">' . htmlspecialchars(implode("\n", $this->value)) . '</textarea>';
        return $editHtml;
    }

    public function getValueDisplay(): string {
        return htmlspecialchars(implode(", ", $this->value));
    }
}

// Specific class for the Accepted Insurance attribute
class PropHandler_Insurance extends PropGeneric_MultiTextBox {

    public function __construct($value) {
        parent::__construct($value, 'Accepted Insurance');
    }

    public function validate(): bool {
        // Specific validation for insurances
        return is_array($this->value) && !empty($this->value);
    }
}

// Usage example
$attributeLgbtq = new PropHandler_LgbtqIdentified(true);
echo $attributeLgbtq->getEdit();
echo $attributeLgbtq->getValueDisplay();

$attributeCategory = new PropHandler_Category("Specialist");
echo $attributeCategory->getEdit();
echo $attributeCategory->getValueDisplay();

$attributeInsurance = new PropHandler_Insurance(["Aetna", "Blue Cross"]);
echo $attributeInsurance->getEdit();
echo $attributeInsurance->getValueDisplay();
`

**Comment by gilbertozioma** [comment]
@CedricTiako Are you working on this issue already?

**Comment by paigejulianne** [comment]
@CedricTiako @gilbertozioma Okay, could this be implemented as a Laravel Livewire component?  https://livewire.laravel.com/docs/components

Basically, I want to return bits of HTML code up to the view based on the property.  So, for example, while 'InsuranceAccepted' might be a row of checkboxes, 'Speciality' might just be an input box.  

That help at all?

**Comment by paigejulianne** [comment]
Perhaps I'm overcomplicating this and making it too inflexible.  Perhaps what it needs to be is another in the table in the table that may define the property and it's metaproperties.  For example, in PHP notation, we might have this in the database:

["shortName" => 'Insurance',
 "displayName" => 'Insurances Accepted',
 'type' => 'checkboxes',
 'multipleAllowed' => true,
 'availableOptions' => [
    "Aetna", "Cigna", "Blue Cross"
 ]
]

Basically, it needs to have a flexible user interface, so no matter what "kind" of property it is, it'll dynamically display on the page.

**Comment by gilbertozioma** [comment]
@paigejulianne, I sent an email to this address tech@genderdirectory.com(mailto:tech@genderdirectory.com) requesting the MySQL database password, but I have not received a response yet. Also, using Laravel Livewire is a great idea for creating a dynamic and flexible user interface that updates in real-time.
